tests/integrationv2/test_early_data.py

```diff
@@ -0,0 +1,397 @@
+
+    server_options = copy.copy(options)
+    server_options.mode = Provider.ServerMode
+
+    s2n_server = managed_process(S2N, server_options)
+    client = managed_process(provider, client_options)
+
+    for results in client.get_results():
+        assert results.exception is None
+        assert results.exit_code == 0
+
+    for results in s2n_server.get_results():
+        assert results.exception is None
+        assert results.exit_code == 0
+        assert to_bytes(S2N_EARLY_DATA_RECV_MARKER) not in results.stdout
+        assert to_bytes(S2N_EARLY_DATA_REJECTED_MARKER) in results.stdout
+        assert DATA_TO_SEND in results.stdout
+
+
+"""
+Test the S2N client attempting to send early data, but the server triggering a hello retry.
+
+We trigger the HRR by configuring the server to only accept curves that the S2N client
+does not send key shares for.
+"""
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", TLS13_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("curve", S2N_HRR_CURVES, ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", ALL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", [Protocols.TLS13], ids=get_parameter_name)
+@pytest.mark.parametrize("provider", SERVER_PROVIDERS, ids=get_parameter_name)
+@pytest.mark.parametrize("early_data_size", [int(MAX_EARLY_DATA/2), int(MAX_EARLY_DATA-1), MAX_EARLY_DATA, 1])
+def test_s2n_client_with_early_data_rejected_via_hrr(managed_process, tmp_path, cipher, curve, protocol, provider, certificate, early_data_size):
+    if provider == S2N:
+        pytest.skip("S2N does not respect ProviderOptions.curve, so does not trigger a retry")
+
+    early_data_file = str(tmp_path / EARLY_DATA_FILE)
+    early_data = get_early_data_bytes(early_data_file, early_data_size)
+
+    options = ProviderOptions(
+        port=next(available_ports),
+        cipher=cipher,
+        curve=curve,
+        protocol=protocol,
+        insecure=True,
+        use_session_ticket=True,
+        reconnect=True,
+    )
+    options.ticket_file = None
+    options.early_data_file = early_data_file
+    options.max_early_data = MAX_EARLY_DATA
+
+    client_options = copy.copy(options)
+    client_options.mode = Provider.ClientMode
+
+    server_options = copy.copy(options)
+    server_options.mode = Provider.ServerMode
+    server_options.key = certificate.key # Required for the initial connection
+    server_options.cert = certificate.cert # Required for the initial connection
+    server_options.reconnects_before_exit = NUM_CONNECTIONS
+
+    server = managed_process(provider, server_options)
+    s2n_client = managed_process(S2N, client_options)
+
+    for results in s2n_client.get_results():
+        assert results.exception is None
+        assert results.exit_code == 0
+        assert results.stdout.count(to_bytes(S2N_EARLY_DATA_REJECTED_MARKER)) == NUM_RESUMES
+
+    for results in server.get_results():
```

Comment:
No nice convenient logging message. S2N also doesn't offer any APIs to retrieve that information, so it'd be a pretty heavy change.

In our [HRR tests](https://github.com/aws/s2n-tls/blob/main/tests/integrationv2/test_hello_retry_requests.py#L129-L130) we look for a specific sequence of bytes that indicates a HRR occurred and count the number of ClientHellos / ServerHellos. Both rely on OpenSSL-specific output, aren't super flexible, and seem too heavy-weight to just verify the setup of a test that isn't specifically testing that we do HRR right. I'd prefer to assume we've tested HRR, know it's working as expected, and focus on verifying the early data behavior.

Reply:
The HRR [marker](https://github.com/aws/s2n-tls/blob/main/tests/integrationv2/test_hello_retry_requests.py#L129-L130) should be sent by all providers if they are performing a HRR. In s2nd/s2nc we dont print the logs detailed enough unfortunately to look for this marker in s2nd/s2nc outputs and rely on the OpenSSL logs for it. We should update the logging that s2nd/c provides in a separate issue and update this test accordingly.

Reply:
Even the HRR marker is tricky to match against, because it's subject to wrapping in OpenSSL's raw message logs so exactly what you need to match against changes. And searching for the right combo of ClientHello/ServerHello OpenSSL debugging messages is also a joy, because those aren't actually logging messages sent/received, they're logging Openssl state machine state, and early data apparently causes the Openssl state machine to sometimes repeat states. If I try to use OpenSSL's output to verify HRR, that will become the bulk of the logic in these integration tests, because it's extremely situation dependent and none of it is reuseable. 

What I could do (in a separate PR) is make s2nc/s2nd print the handshake type via `s2n_connection_get_handshake_type_name`. That would solve the HRR problem, and be a useful signal for other tests too.

Reply:
https://github.com/aws/s2n-tls/issues/2871